src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs

```diff
@@ -451,9 +476,9 @@ private static SourceLocation GetAttributeNameStartLocation(SyntaxTreeNode node)
             if (builder != null)
             {
                 // If the attribute was requested by a tag helper but the corresponding property was not a string,
-                // then treat its value as code. A non-string value can be any C# value so we need to ensure the
-                // SyntaxTreeNode reflects that.
-                if (isBoundNonStringAttribute)
+                // then treat its value as code if it was previously markup. A non-string value can be any C# value so
+                // we need to ensure the SyntaxTreeNode reflects that.
+                if (isBoundNonStringAttribute && builder.Kind == SpanKind.Markup)
```

Comment:
Could go either way on the Razor comments. Today they generate nothing and therefore have no impact but definitely break up the consistency of being in C#. As for metacode: Metacode spans can only occur in invalid Razor blocks within the attributes (errors are created). 

I like the option of enforcing only C# in the attributes and that does flow similar to what #436 is getting at. Will remove the restriction on `.Kind`.